Mathlib/LinearAlgebra/FreeAlgebra.lean

```diff
@@ -45,3 +45,15 @@ theorem rank_eq [CommRing R] [Nontrivial R] :
     Cardinal.lift_umax'.{v,u}, FreeMonoid]
 
 end FreeAlgebra
+
+open Cardinal
+
+theorem Algebra.rank_adjoin_le {R : Type u} {S : Type v} [CommRing R] [Ring S] [Algebra R S]
+    (s : Set S) : Module.rank R (adjoin R s) ≤ max #s ℵ₀ := by
+  rw [adjoin_eq_range_freeAlgebra_lift]
+  cases subsingleton_or_nontrivial R
+  · rw [rank_subsingleton]; exact one_le_aleph0.trans (le_max_right _ _)
+  rw [← lift_le.{max u v}]
+  refine (lift_rank_range_le (FreeAlgebra.lift R ((↑) : s → S)).toLinearMap).trans ?_
+  rw [FreeAlgebra.rank_eq, lift_id'.{v,u}, lift_umax.{v,u}, lift_le, max_comm]
+  exact mk_list_le_max _
```

Comment:
I have results similar to this:

```lean
-- TODO: should be also true with `Semiring A`
-- TODO: move to suitable place
theorem _root_.Algebra.lift_card_adjoin_le
    (R : Type u) {A : Type v} [CommSemiring R] [CommSemiring A] [Algebra R A] (s : Set A) :
    Cardinal.lift.{u} (Cardinal.mk (Algebra.adjoin R s)) ≤
    max (max (Cardinal.lift.{v} (Cardinal.mk R)) (Cardinal.lift.{u} (Cardinal.mk s)))
      Cardinal.aleph0 := by
  rw [Algebra.adjoin_eq_range]
  have h := Cardinal.lift_mk_le_lift_mk_of_surjective
    (MvPolynomial.aeval (R := R) (σ := s) Subtype.val).rangeRestrict_surjective
  rw [Cardinal.lift_umax, Cardinal.lift_id'.{v, u}] at h
  exact h.trans MvPolynomial.cardinal_lift_mk_le_max

-- TODO: move to suitable place
theorem _root_.IntermediateField.lift_card_adjoin_le
    (F : Type u) {E : Type v} [Field F] [Field E] [Algebra F E] (s : Set E) :
    Cardinal.lift.{u} (Cardinal.mk (adjoin F s)) ≤
    max (max (Cardinal.lift.{v} (Cardinal.mk F)) (Cardinal.lift.{u} (Cardinal.mk s)))
      Cardinal.aleph0 := by
  rw [(FractionRing.algEquiv (Algebra.adjoin F s) (adjoin F s)).toEquiv.symm.cardinal_eq,
    FractionRing.card (Algebra.adjoin F s)]
  exact Algebra.lift_card_adjoin_le F s
```

But maybe they should go to another PR.

Reply:
I think `Algebra.lift_card_adjoin_le` should just go above or below your result `Algebra.rank_adjoin_le` (not sure, as it needs to import some files about cardinal). I'm not sure where `IntermediateField.lift_card_adjoin_le` goes.

Reply:
> I think `Algebra.lift_card_adjoin_le` should just go above or below your result `Algebra.rank_adjoin_le` (not sure, as it needs to import some files about cardinal).

Yeah probably. You can use [FreeAlgebra.equivMonoidAlgebraFreeMonoid](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/FreeAlgebra.html#FreeAlgebra.equivMonoidAlgebraFreeMonoid) to prove `Cardinal.mk_freeAlgebra` (`FreeAlgebra.cardinal_eq`) but you still need to import `mk_finsupp` (`mk_list` is already imported).

> I'm not sure where `IntermediateField.lift_card_adjoin_le` goes.

```
import Mathlib
theorem x : True := 
  have := @IntermediateField.adjoin.{0,0}
  have := @MvPolynomial.cardinal_mk_le_max.{0}
  trivial
#find_home x
```
suggests Mathlib.FieldTheory.Cardinality.

Reply:
`Mathlib.FieldTheory.Cardinality` sounds a good file. I'll add them once #18004 is merged. Meanwhile, would you like to split your changes to `Mathlib/LinearAlgebra/FreeAlgebra.lean` as a small PR?

Reply:
What is the status here? Is this now resolved?

Reply:
I think @acmepjz's results haven't been PR'd yet.

Reply:
> What is the status here? Is this now resolved?

No worries. It is just a discussion.